File: algorithms/003_stack/stack.py
class Node():
	def __init__(self, value, prev):
		self.value=value
		self.prev=prev
		
class Stack():
	def __init__(self, top):
		self.top=top
	
	def isEmpty(self):
		if self.top is None:
			return True

	def peek(self):
		return self.top.value

	def push(self, node):
		#check if it's empty <- Sarah's solution
		node.prev=self.top
		self.top=node
		
	def pop(self):
		if self.isEmpty()is None:
			temp=self.top
			self.top=self.top.prev
			return temp

def main():
	new_node = Node (5,None)
	second_node=Node(6,new_node)
	third_node=Node(7,second_node)
	fourth_node=Node(8,third_node)

	new_stack = Stack(new_node)

	new_stack.push(second_node)
	new_stack.push(third_node)
	new_stack.push(fourth_node)
	
	print new_stack.peek()

# Question: if I return temp up in def, how can I print its value vs just calling a function?
	
	new_stack.pop()
	print new_stack.peek()

	

if __name__ == '__main__':
  main()
  #TODO
  #REVERSE A STRING

File: algorithms/003_stack/stack.rb
# sllewelyn

class Stack
  def push(val)
    node = Node.new(val)
    if is_empty?
      @last = node
    else
      node.prev = @last
      @last = node
    end
  end

  def pop
    return if is_empty?
    output = @last
    @last = @last.prev
    output.val
  end

  def is_empty?
    @last.nil?
  end
end

class Node
  attr_accessor :prev, :val

  def initialize(val)
    @val = val
  end
end


# Example reverse a string

stack = Stack.new
str = 'alphabet'
rstr = ''
str.chars.each { |c| stack.push(c) }
while !stack.is_empty?
  rstr = stack.pop
  puts rstr
end

puts rstr
